File: objpascal/step1_read_print.pas
program Mal;

{$H+} // Use AnsiString

Uses sysutils,
     CMem,
     mal_readline,
     mal_types,
     reader,
     printer;

var
    Repl_Env : string = '';
    Line     : string;

// read
function READ(const Str: string) : TMal;
begin
    READ := read_str(Str);
end;

// eval
function EVAL(Ast: TMal; Env: string) : TMal;
begin
    EVAL := Ast;
end;

// print
function PRINT(Exp: TMal) : string;
begin
    PRINT := pr_str(Exp, True);
end;

// repl
function REP(Str: string) : string;
begin
    REP := PRINT(EVAL(READ(Str), Repl_Env));
end;

begin
    while True do
    begin
        try
            Line := _readline('user> ');
            if Line = '' then continue;
            WriteLn(REP(Line))
        except
            On E : MalEOF do Halt(0);
            On E : Exception do
            begin
                WriteLn('Error: ' + E.message);
                WriteLn('Backtrace:');
                WriteLn(GetBacktrace(E));
            end;
        end;
    end;
end.

File: impls/perl6/step1_read_print.pl
use v6;
use lib IO::Path.new($?FILE).dirname;
use reader;
use printer;
use types;

sub read ($str) {
  return read_str($str);
}

sub eval ($ast) {
  return $ast;
}

sub print ($exp) {
  return pr_str($exp, True);
}

sub rep ($str) {
  return print(eval(read($str)));
}

sub MAIN {
  while (my $line = prompt 'user> ').defined {
    say rep($line);
    CATCH {
      when X::MalException { .Str.say }
    }
  }
}
